hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/safemode/ContainerSafeModeRule.java

```diff
@@ -71,83 +84,166 @@ public ContainerSafeModeRule(String ruleName, EventQueue eventQueue,
-    if (maxContainer == 0) {
+    if (ratisMaxContainer == 0) {
+      return 1;
+    }
+    return (ratisContainerWithMinReplicas.doubleValue() / ratisMaxContainer);
+  }
+
+  @VisibleForTesting
+  public synchronized double getCurrentECContainerThreshold() {
+    if (ecMaxContainer == 0) {
+      return 1;
+    }
+    return (ecContainerWithMinReplicas.doubleValue() / ecMaxContainer);
+  }
+
+  public synchronized double getEcMaxContainer() {
+    if (ecMaxContainer == 0) {
       return 1;
     }
-    return (containerWithMinReplicas.doubleValue() / maxContainer);
+    return ecMaxContainer;
+  }
+
+  private synchronized double getRatisMaxContainer() {
+    if (ratisMaxContainer == 0) {
+      return 1;
+    }
+    return ratisMaxContainer;
   }
 
   @Override
   protected synchronized void process(
       NodeRegistrationContainerReport reportsProto) {
+    DatanodeDetails datanodeDetails = reportsProto.getDatanodeDetails();
+    UUID datanodeUUID = datanodeDetails.getUuid();
+    StorageContainerDatanodeProtocolProtos.ContainerReportsProto report = reportsProto.getReport();
+
+    report.getReportsList().forEach(c -> {
+      long containerID = c.getContainerID();
 
-    reportsProto.getReport().getReportsList().forEach(c -> {
-      if (containerMap.containsKey(c.getContainerID())) {
-        if (containerMap.remove(c.getContainerID()) != null) {
-          containerWithMinReplicas.getAndAdd(1);
-          getSafeModeMetrics()
-              .incCurrentContainersWithOneReplicaReportedCount();
+      if (ratisContainerMap.containsKey(containerID)) {
+        ratisContainerDNsMap.computeIfAbsent(containerID, key -> Sets.newHashSet());
+        ratisContainerDNsMap.get(containerID).add(datanodeUUID);
+        if (!reportedConatinerIDSet.contains(containerID)) {
+          Set<UUID> uuids = ratisContainerDNsMap.get(containerID);
+          if (uuids != null && uuids.size() >= 1) {
+            ratisContainerWithMinReplicas.getAndAdd(1);
+            reportedConatinerIDSet.add(containerID);
+            getSafeModeMetrics()
+                .incCurrentContainersWithOneReplicaReportedCount();
+          }
+        }
+      }
+
+      if (ecContainerMap.containsKey(containerID)) {
+        ecContainerDNsMap.computeIfAbsent(containerID, key -> Sets.newHashSet());
+        ecContainerDNsMap.get(containerID).add(datanodeUUID);
+        if (!reportedConatinerIDSet.contains(containerID)) {
+          Set<UUID> uuids = ecContainerDNsMap.get(containerID);
+          ContainerInfo containerInfo = ecContainerMap.get(containerID);
+          ReplicationConfig replicationConfig = containerInfo.getReplicationConfig();
+          if (replicationConfig != null && replicationConfig instanceof ECReplicationConfig) {
+            ECReplicationConfig ecReplicationConfig = (ECReplicationConfig) replicationConfig;
+            int data = ecReplicationConfig.getData();
+            if (uuids != null && uuids.size() > data) {
```

Comment:
For Ratis, just one replica per container is required. So for EC, data number of Datanodes should be sufficient. What do you think?

Reply:
You are right. For EC, the amount of data we have is already sufficient. I will improve the code.